### PR TITLE
php: depend on openldap and resource curl

### LINKED ISSUE #32916: LDAP issues with Mojave PHP7.0+
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [X] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [X] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [X] ran `brew update` and can still reproduce the problem?
- [X] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [X] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?  https://gist.github.com/7b3887ba54a331521857cc27f5e3afef
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew install` commands)

I was trying to run the ldap_connect('ldaps:\\something:port') method and received the HTML page:
localhost didn’t send any data.
ERR_EMPTY_RESPONSE

Looking at the error.log for apache I see this:
Assertion failed: (LDAP_VALID( ld )), function ldap_get_option, file options.c, line 114.
[Thu Oct 11 08:53:17.749420 2018] [core:notice] [pid 1331] AH00052: child pid 2051 exit signal Abort trap (6)

This is with PHP 7.0, 7.1, and 7.2 - I have not tried 5.6 since it isn't used by us.

Even running the ldap_connect() with no parameters produces the same error.

I created a regular docker container using the same exact code and it works fine using the latest version of PHP.

### PR DESCRIPTION
**Pull request by kabel**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Fixes #32916. This ensures we don't have crashes related to mixed crypto or ldap libs. 

While it does bring in a large dependency, it better than trying to work around the problems created by trying to mix system and homebrew libs. curl is built in a manner that would make the features match what the system lib provides.

The patch to `PHP_ADD_INCLUDE`, should have been in the original patch. This was mentioned on the existing [upstream PR](https://github.com/php/php-src/pull/3472#issuecomment-436854897). We just got lucky that macOS's system headers files were compatible with the homebrew library includes.

The patch to `PHP_SETUP_ICONV` is to support the system iconv, by means of searching for `.tbd`. This has been [reported upstream ](https://github.com/php/php-src/pull/3616)in a more generic manner.

### CODE REVIEW

**Comment by SMillerDev** [comment]
@kabel can you drop the revision? We'll merge it with the version bump then.

**Comment by kabel** [comment]
Sure. Revision dropped.

File: Formula/php.rb
Comment (javian) [inline Formula/php.rb]: Can you explain why this is removed ? Does it relate to the added patch block ?

File: Formula/php.rb
Comment (javian) [inline Formula/php.rb]: Can you elaborate on the purpose of this patch ? It looks like MacOS uses extensions that isn't picked up by the configuration script which should be fixed upstream - are there any PRs including this code ?

File: Formula/php.rb
Comment (javian) [inline Formula/php.rb]: Due to the dependency on `boost` we should try and avoid depending on `nghttp2` if possible but see comment below about curl.

File: Formula/php.rb
Comment (javian) [inline Formula/php.rb]: We were discussing to create a `curl-openssl` Formula instead of resourcing it especially since multiple versions of PHP will use it. We would want to avoid managing the same resource multiple times if it can be avoided. I was starting to look at this but never got to a point where I was happy but I was imagining a formula that's completely dependent on other homebrew formula for its dependencies. Perhaps even resourcing nghttp2 in it to get around the boost dependency to be able to support http2 but I'm not sure what the other maintainers think of that.

**Review by javian — CHANGES_REQUESTED** [review]

File: Formula/php.rb
Comment (javian) [reply Formula/php.rb]: Sorry I missed a line in the PR so that explains it.

File: Formula/php.rb
Comment (javian) [reply Formula/php.rb]: I saw your notice about `Patch changes are to ensure a build can work with system ldap or iconv.` - if we depend on the `openldap` Formula why do we need to patch it so that it builds with system ldap ?

File: Formula/php.rb
Comment (kabel) [inline Formula/php.rb]: This through the end of the file _could_ be left off it we're not interested in trying to support system LDAP.

File: Formula/php.rb
Comment (kabel) [reply Formula/php.rb]: We could leave it out, but then it wouldn't be feature compatible with what system curl provides. Perhaps in addition to a new formula for `curl-openssl`, we add a formula for `nghttp2-libonly` that only has the `--enable-lib-only` configure flag (no dependencies outside of pkg-config).

File: Formula/php.rb
Comment (kabel) [reply Formula/php.rb]: I agree that having a new formula would be a more ideal solution. If there was some kind of online discussion, it would be nice to link to it so we can refer to it later when people land here or on the linked issue.

**Comment by javian** [comment]
It currently fails with
```
04:28:15 Error: php: failed
04:28:15 An exception occurred within a child process:
04:28:15   Test::Unit::AssertionFailedError: Zend OPCache extension not loaded.
```

File: Formula/curl-openssl.rb
Comment (javian) [inline Formula/curl-openssl.rb]: Wouldn't we need to specifiy the ldap library here like `--with-ldap-lib=#{Formula["openldap"].opt_prefix}` or does it figure that out by itself ?

File: Formula/php.rb
Comment (javian) [inline Formula/php.rb]: I'm wondering if we should make this an inreplace line instead of a patch, might be easier to manage over time.

File: Formula/curl-openssl.rb
Comment (kabel) [reply Formula/curl-openssl.rb]: pkg-config doing its job, it gets detected from there.

File: Formula/php.rb
Comment (kabel) [reply Formula/php.rb]: You could perhaps handle it with inreplace, but we still have to support both `.dylib` and `.tbd` suffixes exposed by supported macOS platforms.

File: Formula/nghttp2-libonly.rb
Comment (kabel) [reply Formula/nghttp2-libonly.rb]: Right, arguable this doesn't fit the common meaning of `:versioned_formula`, however the displayed explanation is still accurate, `this is an alternate version of another formula`. It a different version, just not a different version as it relates to release version numbers.

File: Formula/curl-openssl.rb
Comment (kabel) [reply Formula/curl-openssl.rb]: I'd rather have a message that hints this is a different version of `curl` vs a formula that isn't linked because it is also provided by macOS. Same sort of logic for `keg_only` reason for `nghtt2-libonly`.

I guess this would be the start of a discussion about how the maintainers would like to handle these kinds of formulae now that formula options are discouraged.

**Comment by kabel** [comment]
I don't know why the `postinstall` step is failing now. That's one method I didn't change. The test then fails because it requires postinstall to be successful in deploying the configuration for `opcache.ini`. The bottle is fine. I downloaded the artifact and ran the tests locally and had no issues. Could there be a jenkins issue?

File: Formula/nghttp2-libonly.rb
Comment (javian) [inline Formula/nghttp2-libonly.rb]: We'll not include this formula for now so depending on `nghttp2` would be the way forward despite the boost dep. Can you switch it back ?

Comment (kabel) [reply Formula/nghttp2-libonly.rb]: Sure. It was getting pulled in by httpd anyway, although it too could benefit from a lighter nghttp2.

File: Formula/php.rb
Comment (javian) [reply Formula/php.rb]: But `/usr/include/iconv.h` exists in Mojave at least since they restored `/usr/include` in 10.14.1 .

**Comment by javian** [comment]
It might be a jenkins issue since we've seen issues in the system. Lets wait for that to get resolved and then we can try again.

File: Formula/php.rb
Comment (kabel) [reply Formula/php.rb]: If they truly did "about face" on that one it would be fine, but my upgraded mojave machine does not have a `/usr/include`. I still think it requires an additional package to be installed to get them. I have XCode, but not CLT on 10.14.1.

Comment (javian) [reply Formula/php.rb]: And I have CLT but not XCode so it might be clt related then since I've not installed anything in addition to make this work.

File: Formula/curl-openssl.rb
Comment (javian) [reply Formula/curl-openssl.rb]: I don't think we will be seeing a lot of these types of formulas so this is going to be an exception. Technically its not a different version in the same sense `versioned_formula` is normally used though since this should be at the same level as the regular curl formula. I guess it doesn't really make a big difference since the output is the same with the exception of one string.

File: Formula/php.rb
Comment (mizunashi-mana) [reply Formula/php.rb]: Apple distributed CLT of High Sierra by software update for Mojave.
It removes `/Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg` and restores `/usr/include` .

Comment (javian) [reply Formula/php.rb]: I just found out that supposedly there might be a bug that did just that so it doesn't seem to happen on all systems. Do you have any links to any documented information about it @mizunashi-mana ?

Comment (javian) [reply Formula/php.rb]: They seem to have taken the affected package out of distribution now so it shouldn't happen anymore.

Comment (mizunashi-mana) [reply Formula/php.rb]: @javian Actually, I'm waiting official announce and cannot find any documents.  However, the summary said such when I updated.  We can see it by System Information > Software > Install.  The latest CLT update shows `Command Line Tools (macOS High Sierra version 10.13) for Xcode` with version 10.1 .

File: Formula/php.rb
Comment (javian) [reply Formula/php.rb]: Could complement the information about the patches in the initial post and include the links there too ? Just so that its easy to find the explanation if we would need to trace it back at a later date

Comment (kabel) [reply Formula/php.rb]: I'm not sure what you're asking for here. Are you suggesting additional URLs be added around the patch line at line 42?

Comment (javian) [reply Formula/php.rb]: No I meant the github post so that we don't have to search through the patch comments =)

**Review by javian — APPROVED** [review]

**Comment by javian** [comment]
@fxcoudert I'm feeling content with the changes, can you have a look at the changes made ? Instead of patching the source code we could rely on the `iconv` formula but I'm not sure which way to go here.

**Review by SMillerDev — APPROVED** [review]

**Comment by javian** [comment]
If no one has merged this before I get home from work today I will merge it then.

**Comment by fxcoudert** [comment]
@javian merged!

**Pull request closed without merging**
